### ISSUE TITLE
SQLite exception with multiple threads

### ISSUE THREAD

**Issue by de-code** [deckar01/ratelimit#4]
Hi,

I am getting the following exception trying to call the decorated function from multiple threads:

```text
ProgrammingError                          Traceback (most recent call last)
/opt/conda/lib/python3.7/site-packages/ratelimit/decorators.py in wrapper(*args, **kargs)
    104                     with self.database:
--> 105                         self.database.execute("BEGIN TRANSACTION")
    106                         # If the number of attempts to call the function exceeds the

ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140303808173888 and this is thread id 140302522234624.

During handling of the above exception, another exception occurred:
```

That is without passing in any `storage` parameter, i.e. it would use the default `:memory:`.

I was using going to use your fork for the "sliding log", not for the persistence.

**Comment by deckar01** [comment]
Sorry I didn't see this sooner. I'm not getting notifications for this project for some reason. I will look into that. I stumbled upon [your upstream comment](https://github.com/tomasbasham/ratelimit/issues/47#issuecomment-691177161) and was about to ask you to open this issue. You put a hyphen in my handle, so it never pinged me from there either. 😄 

Edit: Fixed. The default notification setting for this fork was mentions only.

---

[This SO answer](https://stackoverflow.com/a/24708173) seems to indicate that the new URI format (python>=3.4) supports sharing the memory between threads. `file:ratelimit?mode=memory&cache=shared`.

**Comment by deckar01** [comment]
This should be resolved in 3.0.1. Can you test your use case against this version?

https://github.com/deckar01/ratelimit/releases/tag/3.0.1

**Comment by raratiru** [comment]
The following code works great when calling each function using different REPLs:

```python
@limits(calls=4000, period=60, storage="limit.db")
def api_limit():
    pass

def func1():
    # Imported and called from REPL1
    count = 1
    while True:
        api_limit()
        print(count)
        count += 1

def func2():
    # Imported and called from REPL2
    count = 1
    while True:
        api_limit()
        print(count)
        count += 1
```

When the errors are raised, the sum of the last printed numbers in each REPL, equals 4001.
